src/js/forms.js

```diff
@@ -1,201 +1,99 @@
-
-		this.opts = {
-			testEvent,
-			applyValidState
+	/**
+	* Class constructor.
+	* @param {HTMLElement} [formElement] - The form element in the DOM
+	* @param {Object} [options={}] - An options object for configuring the form
+	*/
+	constructor(formElement, options) {
+		this.formEl = formElement;
+		this.inputFields = Array.from(formElement.querySelectorAll('input', 'select', 'textarea'));
+
+		this.opts = Object.assign({
+			useBrowserValidation: true
+		}, options);
+
+		this.class = {
+			invalid: 'o-forms-input--invalid',
+			valid: 'o-forms-input--valid'
 		};
 
-		if (declaredTestEvent) {
-			this.opts.testEvent = declaredTestEvent;
-		}
-
-		if (declaredApplyValidState) {
-			this.opts.applyValidState = declaredApplyValidState === 'true';
-		}
-
-		// o-forms should only be registered against a <form>
-		// element. If not, return to prevent errors
-		if (!(this.FormEl instanceof HTMLFormElement)) {
-			this.destroy();
-			return;
-		}
-
-		// Add the event listeners
-		this.listeners();
-	}
-
-	listeners() {
-		// Setup checkbox toggle functionality.
-		// We listen for clicks as IE does not fire onchange until onblur.
-		this.FormEl.addEventListener('click', this.handleClickEvent.bind(this), false);
-
-		// Setup validity checks.
-		if (this.opts.testEvent === 'submit') {
-			// Safari reports the validity state, but doesn't
-			// prevent form submits, so this listens to submits and
-			// checks the inputs are valid before submission.
-			this.FormEl.addEventListener('submit', this.validateForm.bind(this), false);
-
-			// All other browsers will report each item invalid on
-			// submit and prevent a form submission.
-			this.findInputs().forEach(input => {
-				input.addEventListener('invalid', this.handleInvalidEvent.bind(this), false);
-			});
+		if (!this.opts.useBrowserValidation) {
+			this.formEl.setAttribute('novalidate', true);
+			this.formEl.addEventListener('submit', this);
 		} else {
-			this.findInputs().forEach((input) => {
-				input.addEventListener('blur', this.handleBlurEvent.bind(this), false);
-			});
+			this.formEl.removeAttribute('novalidate');
+			let submit = this.formEl.querySelector('input[type=submit]');
+			submit.addEventListener('click', this);
+			submit.addEventListener('keydown', this);
 		}
-	}
-
-	validateForm(event) {
-		event.preventDefault();
 
-		const checkedInputs = this.findInputs().map(input => this.validateInput(input));
-
-		if (checkedInputs.some((val) => val === false)) {
-			return;
-		}
-
-		// Complete the form submission
-		event.target.submit();
+		this.inputFields.forEach(input => {
+			input.addEventListener('blur', this);
+			input.addEventListener('input', this);
+		});
 	}
 
-	validateInput(input) {
-		if (input.checkValidity() === false) {
-			this.invalidInput(input);
-			return false;
+	/**
+	 * Event Handler
+	 * @param {Object} event - The event emitted by element/window interactions
+	 */
+	handleEvent(e) {
```

Comment:
Not sure where to put this comment, but if you enter invalid input in one of the demo date fields then shift+tab backwards from year to month, the error state goes away

Reply:
I believe I've fixed it!